projects/control-service/projects/gpu_management/README.md

```diff
@@ -0,0 +1,103 @@
+
+
+Next we need to decide on the constraints of our system.
+
+
+#### Constraint 1:
+A job can only exist on a single machine.
+It would be senseless to deploy a job to more than one machine.
+
+|                               | Job 1 | Job 2 | Job 3 | Job 4 | Job 5 | Job 6 |
+|-------------------------------|--|--|--|--|--|--|
+| Machine 1                     |  ✅(1) | ❌(0) |  ✅(1) | ❌(0) | ❌(0) | ❌(0) |
+| Machine 2                     | ❌(0) | ❌(0) | ❌(0) |  ✅(1) | ❌(0) | ❌(0) |
+| Machine 3                     | ❌(0) | ❌(0) | ❌(0) | ❌(0) |  ✅(1) | ❌(0) |
+| Constaint 1:<br/> Column <= 1 |   1 |  0 | 1 | 1 | 0 | 0 |
+
+```
+job_1___machine_1 + job_1___machine_2 + job_1___machine_3 <= 1
+job_2___machine_1 + job_2___machine_2 + job_2___machine_3 <= 1
+job_3___machine_1 + job_3___machine_2 + job_3___machine_3 <= 1
+job_4___machine_1 + job_4___machine_2 + job_4___machine_3 <= 1
+job_5___machine_1 + job_5___machine_2 + job_5___machine_3 <= 1
+job_6___machine_1 + job_6___machine_2 + job_6___machine_3 <= 1
+```
+
+
+#### Constraint 2
+The total number of jobs on a machine must be less than the amount of resources available on that machine.
+For this we are going to multiply the variable by the amount of resources in consumes.
+For this example we will say:
+1. All machines have 10 devices.
+2. Job 1,2,3 need 4 devices and job 4,5,6 need 6 devices
+
+
+|               | Job 1  | Job 2 | Job 3 | Job 4  | Job 5 | Job 6 | constraint 2:<rb/> Row must be less than number of devices |
+|---------------|--------|--|--|--------|--|--|-----------------------------------------------------------|
+| Machine 1 (devices == 10) | ✅(1*4) | ❌(0*4) |  ✅(1*4) | ❌(0*6) | ❌(0*6) | ❌(0*6) | 8 devices consumed                                        |
+| Machine 2 (devices == 10)    | ❌(0*4)   | ❌(0*4) | ❌(0*4) | ✅(1*6)   | ❌(0*6) | ❌(0*6) | 6 devices consumed                                        |
+| Machine 3 (devices == 10)    | ❌(0*4)   | ❌(0*4) | ❌(0*4) | ❌(0*6)   |  ✅(1*6) | ❌(0*6) | 6 devices consumed                                                         |
+
+
+
+```
+job_1___machine_1*job_1_required_resources + job_2___machine_1*job_2_required_resources + job_3___machine_1*job_3_required_resources + job_4___machine_1*job_4_required_resources + job_5___machine_1*job_5_required_resources + job_6___machine_1*job_6_required_resources <= devices_on_machines_1
+job_1___machine_2*job_1_required_resources + job_2___machine_2*job_2_required_resources + job_3___machine_2*job_3_required_resources + job_4___machine_2*job_4_required_resources + job_5___machine_2*job_5_required_resources + job_6___machine_2*job_6_required_resources <= devices_on_machines_2
+job_1___machine_3*job_1_required_resources + job_2___machine_3*job_2_required_resources + job_3___machine_3*job_3_required_resources + job_4___machine_3*job_4_required_resources + job_5___machine_3*job_5_required_resources + job_6___machine_3*job_6_required_resources <= devices_on_machines_3
+```
+
+
```

Comment:
Realistically we would probably need more constraints. 

- Job Compatibility with Machine as some jobs may need certain machines due to GPU capabilities  (GPU model) or versions requirements. I don't know how pervasive this is as an issue though
- Job priority . Not all jobs are equal. 

We are also considering just number of GPUs. I wonder if GPU memory is not important resource as well.

Reply:
Also it would be cool (not sure if needed though) if we allow users to define their own constraints in some DSL.

Reply:
@antoniivanov I updated the documentation. 
All constraints were not added yesterday. 
They are all written down now. 

The extra documentation describes how priority works. 



For the moment we are considering that all GPUs are equivalent.
Later we can support heterogeneous layouts but for the moment I think the current number of constraints is more than enough?